code/modules/mob/living/carbon/superior_animal/defense.dm

```diff
@@ -0,0 +1,286 @@
+	if (stat != DEAD)
+		target_mob = null
+		stance = initial(stance)
+		stop_automated_movement = initial(stop_automated_movement)
+		walk(src, 0)
+
+		density = 0
+		layer = LYING_MOB_LAYER
+
+	. = ..()
+
+/mob/living/carbon/superior_animal/rejuvenate()
+	density = initial(density)
+	layer = initial(layer)
+
+	. = ..()
+
+/mob/living/carbon/superior_animal/pl_effects()
+	. = ..()
+	adjustToxLoss(2)
+
+/mob/living/carbon/superior_animal/get_cold_protection(var/temperature)
+	return cold_protection
+
+/mob/living/carbon/superior_animal/get_heat_protection(var/temperature)
+	return heat_protection
+
+/mob/living/carbon/superior_animal/handle_environment(var/datum/gas_mixture/environment)
+	if(!environment)
+		return
+
+	if (!contaminant_immunity)
+		for(var/g in environment.gas)
+			if(gas_data.flags[g] & XGM_GAS_CONTAMINANT && environment.gas[g] > gas_data.overlay_limit[g] + 1)
+				pl_effects()
+				break
+
+	if(istype(get_turf(src), /turf/space))
+		if (bodytemperature > 1)
+			bodytemperature = max(1,bodytemperature - 10*(1-get_cold_protection(0)))
+
+		if (min_air_pressure > 0)
+			adjustBruteLoss(2)
+	else
+		var/loc_temp = T0C
+		var/loc_pressure = 0
+		if(istype(loc, /obj/mecha))
+			var/obj/mecha/M = loc
+			loc_temp =  M.return_temperature()
+			loc_pressure =  M.return_pressure()
+		else if(istype(loc, /obj/machinery/atmospherics/unary/cryo_cell))
+			loc_temp = loc:air_contents.temperature
+			loc_pressure = loc:air_contents.return_pressure()
+		else
+			loc_temp = environment.temperature
+			loc_pressure = environment.return_pressure()
+
+		//Body temperature adjusts depending on surrounding atmosphere based on your thermal protection (convection)
+		var/temp_adj = 0
+		if(loc_temp < bodytemperature) //Place is colder than we are
+			var/thermal_protection = get_cold_protection(loc_temp) //0 to 1 value, which corresponds to the percentage of protection
+			if(thermal_protection < 1)
+				temp_adj = (1-thermal_protection) * ((loc_temp - bodytemperature) / BODYTEMP_COLD_DIVISOR) //this will be negative
+		else if (loc_temp > bodytemperature) //Place is hotter than we are
+			var/thermal_protection = get_heat_protection(loc_temp) //0 to 1 value, which corresponds to the percentage of protection
+			if(thermal_protection < 1)
+				temp_adj = (1-thermal_protection) * ((loc_temp - bodytemperature) / BODYTEMP_HEAT_DIVISOR)
+
+		var/relative_density = environment.total_moles / MOLES_CELLSTANDARD
+		bodytemperature += between(BODYTEMP_COOLING_MAX, temp_adj*relative_density, BODYTEMP_HEATING_MAX)
+
+		if ((loc_pressure < min_air_pressure) || (loc_pressure > max_air_pressure))
+			adjustBruteLoss(2)
+
+	if (overkill_dust && (getFireLoss() >= maxHealth*2))
+		if (bodytemperature > max_bodytemperature*2)
+			dust()
+			return
+
+	if ((bodytemperature > max_bodytemperature) || (bodytemperature < min_bodytemperature))
+		adjustFireLoss(5)
+		updatehealth()
+
+/mob/living/carbon/superior_animal/handle_breath(datum/gas_mixture/breath)
```

Comment:
breathing code, oh boy

Are there any noticeable performance impacts from this stuff?
Perhaps as an optimisation, you could give nonhuman mobs a longer interval between breath checks. IIRC humans check every life ticks

Reply:
I've tried a hundred nurse spiders and checked profiler, it was much less demanding than the constant icon update cells in apcs were doing which nobody noticed.